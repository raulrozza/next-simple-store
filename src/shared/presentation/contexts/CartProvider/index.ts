export { default as useAddToCart } from './hooks/useAddToCart';
export { default as useCartMetaValue } from './hooks/useCartMetaValue';
export { default as useCartValue } from './hooks/useCartValue';
