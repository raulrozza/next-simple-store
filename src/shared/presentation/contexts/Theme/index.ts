export { default as useTheme } from './hooks/useTheme';
export { default as Theme } from './implementations/StyledComponentTheme';
