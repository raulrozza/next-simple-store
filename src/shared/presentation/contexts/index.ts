export * from './CartProvider';
export * from './ThemeProvider';
