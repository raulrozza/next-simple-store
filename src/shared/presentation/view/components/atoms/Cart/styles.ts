import { HoverCardContent, HoverCardTrigger } from '@radix-ui/react-hover-card';
import styled, { css } from 'styled-components';

export const Trigger = styled(HoverCardTrigger)`
    ${({ theme }) => css`
        background-color: ${theme.palette.primary['800']};
        color: ${theme.palette.gray['0']};
        border-radius: 50%;
        position: relative;

        width: ${theme.layout.size(3)};
        height: ${theme.layout.size(3)};

        display: flex;
        justify-content: center;
        align-items: center;

        font-size: ${theme.typography.sizes.md};

        strong {
            display: flex;
            justify-content: center;
            align-items: center;
            position: absolute;
            right: -${theme.layout.spacing(1)};
            top: -${theme.layout.spacing(1)};

            background-color: ${theme.palette.secondary['800']};
            padding: ${theme.layout.spacing(1)};
            width: ${theme.layout.size(1.5)};
            height: ${theme.layout.size(1.5)};

            border-radius: 50%;
            font-size: ${theme.typography.sizes.xs};
        }
    `}
`;

export const Content = styled(HoverCardContent)`
    ${({ theme }) => css`
        display: flex;
        flex-direction: column;
        align-items: center;
        gap: ${theme.layout.spacing(1)};

        background-color: ${theme.palette.gray['0']};
        border: 1px solid ${theme.palette.secondary['600']};
        border-radius: ${theme.layout.borderRadius.md};
        box-shadow: 0 0 10px 3px #00000029;

        padding: ${theme.layout.spacing(2)};

        font-size: ${theme.typography.sizes.xs};

        strong {
            color: ${theme.palette.primary['1000']};
        }

        a {
            font-size: ${theme.typography.sizes.xs};
            color: ${theme.palette.secondary['1000']};

            &:hover {
                text-decoration: underline;
                color: ${theme.palette.secondary['800']};
            }
        }
    `}
`;
