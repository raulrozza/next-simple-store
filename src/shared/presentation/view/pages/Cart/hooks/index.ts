export { default as useCartController } from './useCartController';
