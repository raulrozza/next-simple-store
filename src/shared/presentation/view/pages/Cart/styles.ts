import styled, { css } from 'styled-components';

import { MAX_SCREEN_WIDTH } from '@/shared/presentation/constants/screen';

export const Container = styled.div`
    display: flex;
    justify-content: center;
`;

export const Content = styled.main`
    ${({ theme }) => css`
        width: 100%;
        max-width: ${theme.layout.size(MAX_SCREEN_WIDTH)};
        margin: ${theme.layout.size(1)};
    `}
`;

export const List = styled.ul`
    ${({ theme }) => css`
        background-color: ${theme.palette.secondary['600']};
        padding: ${theme.layout.spacing(2)};
        border-radius: ${theme.layout.borderRadius.md};

        display: flex;
        flex-direction: column;
        gap: ${theme.layout.spacing(2)};
    `}
`;
