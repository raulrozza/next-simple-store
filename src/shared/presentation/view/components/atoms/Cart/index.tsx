import { FC, useState } from 'react';

import { Root } from '@radix-ui/react-hover-card';
import Link from 'next/link';
import { BsFillCartFill } from 'react-icons/bs';

import { useCartMetaValue } from '@/shared/presentation/contexts';

import { Content, Trigger } from './styles';

const Cart: FC = () => {
    const { quantity, total } = useCartMetaValue();
    const [open, setOpen] = useState(false);

    return (
        <Root open={open} onOpenChange={setOpen} openDelay={0} closeDelay={0}>
            <Trigger onClick={() => setOpen(true)}>
                <BsFillCartFill />

                <strong>{quantity}</strong>
            </Trigger>

            <Content>
                <p>You currently have {quantity} products in your cart.</p>

                <p>
                    Total: <strong>${total.toFixed(2)}</strong>
                </p>

                <Link href="/cart">See my cart</Link>
            </Content>
        </Root>
    );
};

export default Cart;
