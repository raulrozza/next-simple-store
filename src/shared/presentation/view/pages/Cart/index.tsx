import { FC, useMemo } from 'react';

import { FormikProvider, useFormik } from 'formik';

import {
    useCartManager,
    useCartMetaValue,
    useCartValue,
} from '@/shared/presentation/contexts';
import {
    Button,
    CartListItem,
    Spacing,
} from '@/shared/presentation/view/components/atoms';
import { Form } from '@/shared/presentation/view/components/molecules';
import { Menu } from '@/shared/presentation/view/components/organisms';

import { useCartController } from './hooks';
import {
    Container,
    Content,
    FinishOrderContainer,
    List,
    ListFooter,
} from './styles';

const Cart: FC = () => {
    const cart = useCartValue();
    const meta = useCartMetaValue();
    const { clear } = useCartManager();
    const { customers } = useCartController();
    const formik = useFormik({
        initialValues: {
            customer: '',
            installments: 1,
        },
        onSubmit: console.log,
    });

    const installmentOptions = useMemo(() => {
        const selectedCustomer = formik.values.customer;

        if (!selectedCustomer) return [];

        const customer = customers.find(({ id }) => id === selectedCustomer);

        if (!customer) return [];

        return Array.from(
            { length: customer.installmentLimit },
            (_, index) => ({
                value: index + 1,
                label: `${index + 1}x`,
            }),
        );
    }, [customers, formik.values.customer]);

    return (
        <section>
            <Menu activeItem="/orders" />

            <Container>
                <Content>
                    <h1>Cart</h1>

                    <Spacing size={3} />

                    {cart.length ? (
                        <List>
                            {cart.map(entry => (
                                <CartListItem
                                    key={entry.product.id}
                                    {...entry}
                                />
                            ))}

                            <ListFooter>
                                <div>
                                    <p>
                                        You have {meta.quantity} products in
                                        your cart.
                                    </p>

                                    <p>Total: ${meta.total.toFixed(2)}</p>
                                </div>

                                <Spacing size={2} />

                                <Button variant="primary" onClick={clear}>
                                    Remove all items
                                </Button>
                            </ListFooter>
                        </List>
                    ) : (
                        <p>No items added yet.</p>
                    )}

                    <Spacing size={3} />

                    <h2>Finish order</h2>

                    <FormikProvider value={formik}>
                        <FinishOrderContainer>
                            <div>
                                <Form
                                    fields={[
                                        {
                                            type: 'select',
                                            name: 'customer',
                                            placeholder: 'Select a customer',
                                            options: customers.map(
                                                customer => ({
                                                    label: customer.name,
                                                    value: customer.id,
                                                }),
                                            ),
                                        },
                                        {
                                            type: 'select',
                                            name: 'installments',
                                            placeholder:
                                                'Number of installments',
                                            options: installmentOptions,
                                            disabled:
                                                !installmentOptions.length,
                                        },
                                    ]}
                                    button={{
                                        text: 'Submit order',
                                    }}
                                />
                            </div>

                            <div>opa </div>
                        </FinishOrderContainer>
                    </FormikProvider>
                </Content>
            </Container>
        </section>
    );
};

export default Cart;
