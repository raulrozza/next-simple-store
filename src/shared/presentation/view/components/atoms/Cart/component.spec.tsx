import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ThemeProvider } from '@/shared/presentation/contexts';

import Cart from './index';

jest.mock('@/shared/presentation/contexts/CartProvider', () => ({
    useCartMetaValue: jest.fn(() => ({
        quantity: 5,
        total: 46.5,
    })),
}));

describe('Cart', () => {
    it('should render the component showing the correct quantity and with the content hidden', () => {
        const view = render(<Cart />, {
            wrapper: ThemeProvider,
        });

        expect(screen.getByText('5')).toBeInTheDocument();
        expect(screen.queryByText('See my cart')).not.toBeInTheDocument();
        expect(view).toMatchSnapshot();
    });

    it('should show the content when the trigger is clicked', async () => {
        const user = userEvent.setup();

        render(<Cart />, {
            wrapper: ThemeProvider,
        });

        await user.click(screen.getByText('5'));

        expect(
            screen.getByText('You currently have 5 products in your cart.'),
        ).toBeInTheDocument();
        expect(screen.getByText('Total:')).toBeInTheDocument();
        expect(screen.getByText('$46.50')).toBeInTheDocument();
        expect(screen.getByText('See my cart')).toBeInTheDocument();
    });
});
