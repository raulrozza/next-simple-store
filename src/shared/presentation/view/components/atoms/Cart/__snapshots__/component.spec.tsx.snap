// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart should render the component showing the correct quantity and with the content hidden 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  background-color: #4A93A8;
  color: #FFFFFF;
  border-radius: 50%;
  position: relative;
  width: 3rem;
  height: 3rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.5rem;
}

.c0 strong {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: absolute;
  right: -0.25rem;
  top: -0.25rem;
  background-color: #653C48;
  padding: 0.25rem;
  width: 1.5rem;
  height: 1.5rem;
  border-radius: 50%;
  font-size: 0.75rem;
}

<body>
    <div>
      <a
        class="c0"
        data-state="closed"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 1.5A.5.5 0 0 1 .5 1H2a.5.5 0 0 1 .485.379L2.89 3H14.5a.5.5 0 0 1 .491.592l-1.5 8A.5.5 0 0 1 13 12H4a.5.5 0 0 1-.491-.408L2.01 3.607 1.61 2H.5a.5.5 0 0 1-.5-.5zM5 12a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm7 0a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm-7 1a1 1 0 1 1 0 2 1 1 0 0 1 0-2zm7 0a1 1 0 1 1 0 2 1 1 0 0 1 0-2z"
          />
        </svg>
        <strong>
          5
        </strong>
      </a>
    </div>
  </body>,
  "container": .c0 {
  background-color: #4A93A8;
  color: #FFFFFF;
  border-radius: 50%;
  position: relative;
  width: 3rem;
  height: 3rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.5rem;
}

.c0 strong {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: absolute;
  right: -0.25rem;
  top: -0.25rem;
  background-color: #653C48;
  padding: 0.25rem;
  width: 1.5rem;
  height: 1.5rem;
  border-radius: 50%;
  font-size: 0.75rem;
}

<div>
    <a
      class="c0"
      data-state="closed"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 1.5A.5.5 0 0 1 .5 1H2a.5.5 0 0 1 .485.379L2.89 3H14.5a.5.5 0 0 1 .491.592l-1.5 8A.5.5 0 0 1 13 12H4a.5.5 0 0 1-.491-.408L2.01 3.607 1.61 2H.5a.5.5 0 0 1-.5-.5zM5 12a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm7 0a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm-7 1a1 1 0 1 1 0 2 1 1 0 0 1 0-2zm7 0a1 1 0 1 1 0 2 1 1 0 0 1 0-2z"
        />
      </svg>
      <strong>
        5
      </strong>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
